dill/_shims.py

```diff
@@ -79,62 +92,66 @@ def __reduce__(self):
         return self.__reduce__()
 
-class _CallableShimMixin(object):
-    # A version of Shim for functions. Used to trick pickler.save_reduce into
-    # thinking that Shim objects of functions are themselves meaningful functions.
+class _CallableReduce(Reduce):
+    # A version of Reduce for functions. Used to trick pickler.save_reduce into
+    # thinking that Reduce objects of functions are themselves meaningful functions.
     def __call__(self, *args, **kwargs):
         reduction = self.__reduce__()
         func = reduction[0]
         f_args = reduction[1]
         obj = func(*f_args)
         return obj(*args, **kwargs)
 
-class GetAttrShim(Shim):
+__NO_DEFAULT = _dill.Sentinel('Getattr.NO_DEFAULT')
+
+def Getattr(object, name, default=__NO_DEFAULT):
     """
-    A Shim object that represents the getattr operation. When unpickled, the
-    GetAttrShim will access an attribute 'name' of 'object' and return the value
+    A Reduce object that represents the getattr operation. When unpickled, the
+    Getattr will access an attribute 'name' of 'object' and return the value
     stored there. If the attribute doesn't exist, the default value will be
     returned if present.
+
+    The following statements are equivalent:
+
+    Getattr(collections, 'OrderedDict')
+    Getattr(collections, 'spam', None)
+    Getattr(*args)
+
+    Reduce(getattr, (collections, 'OrderedDict'))
+    Reduce(getattr, (collections, 'spam', None))
+    Reduce(getattr, args)
+
+    During unpickling, the first two will result in collections.OrderedDict and
+    None respectively because the first attribute exists and the second one does
+    not, forcing it to use the default value given in the third argument.
     """
-    NO_DEFAULT = _dill.Sentinel('_shims.GetAttrShim.NO_DEFAULT')
-    def __new__(cls, object, name, default=NO_DEFAULT):
-        return Shim.__new__(cls, is_callable=callable(default))
-    def __init__(self, object, name, default=NO_DEFAULT):
-        if object is None:
-            # Use the calling function's module
-            object = sys.modules[inspect.currentframe().f_back.f_globals['__name__']]
-
-        if default is GetAttrShim.NO_DEFAULT:
-            reduction = (getattr, (object, name))
-        else:
-            reduction = (getattr, (object, name, default))
 
-        super(GetAttrShim, self).__init__(reduction)
+    if default is Getattr.NO_DEFAULT:
+        reduction = (getattr, (object, name))
+    else:
+        reduction = (getattr, (object, name, default))
+
+    return Reduce(*reduction, is_callable=callable(default))
 
-        self.object = object
-        self.name = name
-        self.default = default
-    @classmethod
-    def _callable(cls, object, name, default=NO_DEFAULT):
-        return callable(default)
+Getattr.NO_DEFAULT = __NO_DEFAULT
+del __NO_DEFAULT
 
-def _assign_to_dill_module(func):
-    _dill.__dict__[func.__name__] = func
-    return func
+def assign_to(module, name=None):
```

Comment:
What do you think about renaming this to something like: `instantiate_to` or `compile_to` or something using a similarly appropriate technical term?

Reply:
`move_to` is ok.

Reply:
The `@move_to` decorator *must* be paired with the overloading by the `Getattr` later in the module, correct?  It seems like this could be consolidated.  It's worth thinking on that a bit.

Reply:
@anivegesana: Let me know if this should show up as a new issue with the suggested refactor.